Fixes #30368 - Make Candlepin CA file optional

ekohl opened the pull request:
This change makes the Candlepin CA file optional by falling back to the Foreman CA file. The result is a reduced configuration in most deployments.

For Pulp the certificate is now read in the same way as Candlepin. Previously it partly relied on the CA being in the global allowed CA. This may be an issue in some cases, but in the default deployment it isn't. Following the general SSL config makes the configuration more predictable for users. It can also be easier in a containerized setup or on a system where the admin is not allowed to modify the system CA certificates. The example config is now also consistent with reality.

The verify_ssl option is dropped from the ping model. This isn't respected elsewhere and it's misleading to have a valid ping only to have it fail at runtime.

ehelms commented:
Related, and based on some discussion with @jlsherrill , today Pulp communication relies on the the CA being trusted by the system (the whole use of trusted_ca module in the installer) and we should switch to doing this explicitly -- https://github.com/Katello/katello/blob/68ff922e84b66df05f2b7c767c0cd8f2a7328611/app/models/katello/concerns/smart_proxy_extensions.rb#L119

ekohl commented:
I took a stab at this. In doing so, I found 2 settings that were removed but still had references to them. It then also unifies the certificate handling in the certs service. There are a few more changes, documented in the commit message. Please read it and let me hear your thoughts.

ekohl commented on app/lib/katello/resources/registry.rb:
```diff
@@ -33,10 +33,10 @@ def load_class
+              ca_cert_file = ::Cert::Certs.backend_ca_cert_file(:pulp3)
             elsif (container_config = SETTINGS.dig(:katello, :container_image_registry))
               registry_url = container_config[:crane_url]
-              ca_cert_file = container_config[:registry_ca_cert_file]
```
I debated this change. It suggests nobody uses this since the [installer sets `crane_ca_cert_file`](https://github.com/theforeman/puppet-katello/blob/2e15151a6d324149071c4e240bed78b5d977adf6/templates/katello.yaml.erb#L42)

ehelms commented on app/services/cert/certs.rb:
```diff
@@ -24,10 +24,14 @@ def self.ssl_client_key_filename
     end
 
+    def self.backend_ca_cert_file(backend)
+      SETTINGS.dig(:katello, backend, :ca_cert_file) || Setting[:ssl_ca_file]
```
I honestly cannot think of a scenario where we would have or want a different CA cert for Candlepin or Pulp given our architecture. @jlsherrill can you think of any reason to not just set the CA cert used to Setting[:ssl_ca_file] and reduce the potential configuration options?

ekohl replied:
I didn't know how breaking we would be if we dropped it immediately and if that's acceptable.

jlsherrill replied:
unless there is some quirk related with a 3rd party CA i can't think of this should be fine.  I would imagine all sorts of things would not work properly if you started using different CAs for differnet subsystems.

ekohl replied:
There is one case. Pulp 3 can use a different CA than the Smart Proxy. For example, one signed by a publicly trusted CA.

jturel commented:
[test katello]

jturel commented on app/models/katello/ping.rb:
```diff
@@ -226,10 +226,11 @@ def fetch_proxy(capsule_id)
-        options[:ssl_ca_file] = ca_file unless ca_file.nil?
-        options[:verify_ssl] = SETTINGS[:katello][backend][:verify_ssl] if SETTINGS[:katello][backend].key?(:verify_ssl)
+        if (ca_file = ::Cert::Certs.backend_ca_cert_file(backend))
+          options[:ssl_ca_file] = ca_file
```
Was this removed because we don't ever set the `verify_ssl` option for any of the backends?

ekohl replied:
Indeed. It didn't make any sense to have separate logic for ping when the actual resources don't respect the setting.

jturel commented:
One test failure seems relevant

```
test_verify_ueber_cert_no_change – Katello::CertsTest
```

ekohl commented:
Rebased to resolve merge conflict and to get recent tests. I've also split of unrelated work to https://github.com/Katello/katello/pull/9268 since it doesn't impact this PR.

jturel commented:
Haven't tested yet but the test failure definitely seems related

ekohl commented:
Given stabilization is coming soon and my lack of time, I'm afraid we'll have to push this to the next Katello release.

parthaa commented:
@ekohl  is this PR still valid ? Rebase if so

ekohl commented:
I've tried to rebase it, but I'm a bit unclear about the current Pulp 2 vs Pulp 3 status. I still see traces of Runcible, but isn't that Pulp 2-only?

I now think this unifies everything. It does mean the system store isn't really used anymore (unless you don't configure anything I think), but that's probably a good thing.

Perhaps @ehelms and @evgeni could also think about what kind of behavior we would prefer to have from a platform point of view.

ekohl commented on app/lib/katello/http_resource.rb:
```diff
@@ -19,7 +19,7 @@ def initialize(params)
 
     class_attribute :consumer_secret, :consumer_key, :ca_cert_file, :prefix, :site, :default_headers,
-                    :ssl_client_cert, :ssl_client_key
+                    :ssl_client_cert, :ssl_client_key, :ssl_ca_file
```
I now see there already is `ca_cert_file` for this. I don't know what other classes can be passes into `pulp3_ssl_configuration` so I'm looking for guidance here.

ekohl replied:
I included a commit to rename it for consistency with other tools and naming for the other `ssl_` attrs.

ehelms commented on config/katello.yaml.example:
```diff
@@ -17,22 +17,16 @@
     # refers to the apache certificate
-    # (typically /etc/pki/tls/certs/localhost.crt) location that is needed
-    # to connect to pulp over https.
+    # Optional CA file to user to verify HTTPS connections to Candlepin. If not
```
```suggestion
    # Optional CA file to user to verify HTTPS connections to Pulp. If not
```

ehelms reviewed (approved):
Clean re-factoring that helps set up future simplifications.

ekohl commented:
I had hoped @evgeni could test if this resolves his issue in containers before merging.

ehelms commented:
> I had hoped @evgeni could test if this resolves his issue in containers before merging.

I liked the change from a re-factoring stand point. If this doesn't solve his issue, then we can iterate on solving that.

evgeni commented:
https://github.com/theforeman/foreman-quadlet/pull/43 is not failing on SSL errors anymore, so it certainly helped :)
